File: src/test/java/cube8540/oauth/authentication/credentials/oauth/error/OAuth2ExceptionResponseRendererTest.java
package cube8540.oauth.authentication.credentials.oauth.error;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.http.CacheControl;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.server.ServletServerHttpResponse;
import org.springframework.security.oauth2.core.OAuth2Error;
import org.springframework.web.HttpMediaTypeNotSupportedException;
import org.springframework.web.context.request.ServletWebRequest;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@DisplayName("OAuth2 기본 예외 응답 메시지 클래스 테스트")
class OAuth2ExceptionResponseRendererTest {

    @Test
    @DisplayName("지원 되지 않는 미디어 타입으로 클래스 초기화")
    void initializeByNotSupportedMediaType() {
        HttpMessageConverter<Object> messageConverter = makeConverter();

        when(messageConverter.canWrite(any(), any())).thenReturn(false);

        assertThrows(HttpMediaTypeNotSupportedException.class, () -> new OAuth2ExceptionResponseRenderer(messageConverter));
    }

    @Test
    @DisplayName("ResponseEntity의 body가 null 일때 응답 객체에 메시지 작성")
    void writeResponseMessageToResponseObjectWhenResponseEntityBodyIsNull() throws Exception {
        HttpMessageConverter<Object> messageConverter = makeConverter();
        ResponseEntity<OAuth2Error> responseEntity = makeResponseEntity(null);
        OAuth2ExceptionResponseRenderer renderer = new OAuth2ExceptionResponseRenderer(messageConverter);
        ServletWebRequest webRequest = makeWebRequest();

        renderer.rendering(responseEntity, webRequest);
        verify(webRequest.getResponse(), times(1)).setStatus(HttpStatus.UNAUTHORIZED.value());
        verify(webRequest.getResponse(), times(1)).addHeader(HttpHeaders.CACHE_CONTROL, CacheControl.noStore().getHeaderValue());
        verify(webRequest.getResponse(), times(1)).addHeader(HttpHeaders.PRAGMA, "no-cache");
        verify(messageConverter, never()).write(any(), any(), any());
    }

    @Test
    @DisplayName("응답 객체에 메시지 작성")
    void writeResponseMessageToResponseObject() throws Exception {
        OAuth2Error error = mock(OAuth2Error.class);
        HttpMessageConverter<Object> messageConverter = makeConverter();
        ResponseEntity<OAuth2Error> responseEntity = makeResponseEntity(error);
        ArgumentCaptor<ServletServerHttpResponse> responseCaptor = ArgumentCaptor.forClass(ServletServerHttpResponse.class);
        ServletWebRequest webRequest = makeWebRequest();
        OAuth2ExceptionResponseRenderer renderer = new OAuth2ExceptionResponseRenderer(messageConverter);

        renderer.rendering(responseEntity, webRequest);
        verify(webRequest.getResponse(), times(1)).setStatus(HttpStatus.UNAUTHORIZED.value());
        verify(messageConverter, times(1)).write(eq(error), eq(MediaType.ALL), responseCaptor.capture());
        responseCaptor.getValue().getBody(); // header flush
        verify(webRequest.getResponse(), times(1)).addHeader(HttpHeaders.CACHE_CONTROL, CacheControl.noStore().getHeaderValue());
        verify(webRequest.getResponse(), times(1)).addHeader(HttpHeaders.PRAGMA, "no-cache");
        assertEquals(webRequest.getResponse(), responseCaptor.getValue().getServletResponse());
    }

    @SuppressWarnings("unchecked")
    private HttpMessageConverter<Object> makeConverter() {
        HttpMessageConverter<Object> converter = mock(HttpMessageConverter.class);

        when(converter.canWrite(any(), any())).thenReturn(true);

        return converter;
    }

    private HttpHeaders makeHeaders() {
        HttpHeaders header = new HttpHeaders();

        header.setCacheControl(CacheControl.noStore());
        header.setPragma("no-cache");

        return header;
    }

    @SuppressWarnings("unchecked")
    private ResponseEntity<OAuth2Error> makeResponseEntity(OAuth2Error body) {
        ResponseEntity<OAuth2Error>responseEntity = mock(ResponseEntity.class);

        when(responseEntity.getBody()).thenReturn(body);
        when(responseEntity.getStatusCode()).thenReturn(HttpStatus.UNAUTHORIZED);
        when(responseEntity.getHeaders()).thenReturn(makeHeaders());

        return responseEntity;
    }

    private ServletWebRequest makeWebRequest() {
        HttpServletRequest request = mock(HttpServletRequest.class);
        HttpServletResponse response = mock(HttpServletResponse.class);

        return new ServletWebRequest(request, response);
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/error/OAuth2ExceptionResponseRendererTest.kt
package cube8540.oauth.authentication.oauth.error

import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.catchThrowable
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.springframework.http.*
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.server.ServletServerHttpResponse
import org.springframework.security.oauth2.core.OAuth2Error
import org.springframework.web.HttpMediaTypeNotSupportedException
import org.springframework.web.context.request.ServletWebRequest
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class OAuth2ExceptionResponseRendererTest {

    @Nested
    inner class InitializationTest {
        private val messageConverter: HttpMessageConverter<Any> = mockk()

        @Test
        fun `message converter is not supported media type`() {
            every { messageConverter.canWrite(OAuth2Error::class.java, MediaType.APPLICATION_JSON) } returns false

            val thrown = catchThrowable { OAuth2ExceptionResponseRenderer(messageConverter, MediaType.APPLICATION_JSON) }
            assertThat(thrown).isInstanceOf(HttpMediaTypeNotSupportedException::class.java)
        }

        @Test
        fun `message converter is supported media type`() {
            every { messageConverter.canWrite(OAuth2Error::class.java, MediaType.APPLICATION_JSON) } returns true

            assertDoesNotThrow { OAuth2ExceptionResponseRenderer(messageConverter, MediaType.APPLICATION_JSON) }
        }
    }

    @Nested
    inner class RenderingTest {
        private val messageConverter: HttpMessageConverter<Any> = mockk(relaxed = true) {
            every { canWrite(OAuth2Error::class.java, MediaType.APPLICATION_JSON) } returns true
        }
        private val renderer = OAuth2ExceptionResponseRenderer(messageConverter, MediaType.APPLICATION_JSON)

        @Test
        fun `rendering response when body is null`() {
            val httpHeaders = HttpHeaders()
            val responseEntity: ResponseEntity<OAuth2Error> = mockk {
                every { body } returns null
                every { statusCode } returns HttpStatus.BAD_REQUEST
                every { headers } returns httpHeaders
            }
            val httpServletRequest: HttpServletRequest = mockk(relaxed = true)
            val httpServletResponse: HttpServletResponse = mockk(relaxed = true)
            val webRequest: ServletWebRequest = mockk {
                every { request } returns httpServletRequest
                every { response } returns httpServletResponse
            }

            httpHeaders.setCacheControl(CacheControl.noStore())
            httpHeaders.pragma = "no-cache"

            renderer.rendering(responseEntity, webRequest)
            verify { httpServletResponse.status = HttpStatus.BAD_REQUEST.value() }
            verify { httpServletResponse.addHeader(HttpHeaders.CACHE_CONTROL, CacheControl.noStore().headerValue) }
            verify { httpServletResponse.addHeader(HttpHeaders.PRAGMA, "no-cache") }
            verify(exactly = 0) { messageConverter.write(any(), any(), any()) }
        }

        @Test
        fun `rendering response when body is not null`() {
            val responseCaptor = slot<ServletServerHttpResponse>()
            val httpHeaders = HttpHeaders()
            val errorBody: OAuth2Error = mockk()
            val responseEntity: ResponseEntity<OAuth2Error> = mockk {
                every { body } returns errorBody
                every { statusCode } returns HttpStatus.BAD_REQUEST
                every { headers } returns httpHeaders
            }
            val httpServletRequest: HttpServletRequest = mockk(relaxed = true)
            val httpServletResponse: HttpServletResponse = mockk(relaxed = true)
            val webRequest: ServletWebRequest = mockk {
                every { request } returns httpServletRequest
                every { response } returns httpServletResponse
            }

            httpHeaders.setCacheControl(CacheControl.noStore())
            httpHeaders.pragma = "no-cache"

            renderer.rendering(responseEntity, webRequest)
            verify { messageConverter.write(eq(errorBody), eq(MediaType.APPLICATION_JSON), capture(responseCaptor)) }
            assertThat(responseCaptor.isCaptured).isTrue
            responseCaptor.captured.body // header flush
            verify { httpServletResponse.status = HttpStatus.BAD_REQUEST.value() }
            verify { httpServletResponse.addHeader(HttpHeaders.CACHE_CONTROL, CacheControl.noStore().headerValue) }
            verify { httpServletResponse.addHeader(HttpHeaders.PRAGMA, "no-cache") }
            assertThat(httpServletResponse).isEqualTo(responseCaptor.captured.servletResponse)
        }
    }
}